### PR TITLE
Feature expand nn linear

### PR DESCRIPTION
**Pull request by nuniz**
## Proposed changes
1. Added identity and bilinear layers
2. Added a reset_parameters option 
#303 
3. Initialized bias with normal distribution for Torch compatibility.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: python/mlx/nn/layers/linear.py [hunk 1]
```diff
@@ -6,6 +6,21 @@
 from mlx.nn.layers.base import Module
 
 
+class Identity(Module):
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:9]: Could you explain in which cases this will be used? And maybe provide an example?

File: python/mlx/nn/layers/linear.py [hunk 2]
```diff
@@ -26,20 +41,92 @@ class Linear(Module):
 
     def __init__(self, input_dims: int, output_dims: int, bias: bool = True):
         super().__init__()
-        scale = math.sqrt(1 / input_dims)
+        self.input_dims = input_dims
+        self.output_dims = output_dims
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:45]: You don't need to record these.  You can use `self.weight.shape` below.

File: python/mlx/nn/layers/linear.py [hunk 3]
```diff
@@ -26,20 +41,92 @@ class Linear(Module):
...
+
+
+class Bilinear(Module):
+    r"""Applies a bilinear transformation to the input.
+
+    Concretely:
+
+    .. math::
+
+        y = input1^\top W input2 + b
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:83]: x_1 and x_2 in the math will read more cleanly.

File: python/mlx/nn/layers/linear.py [hunk 4]
```diff
@@ -26,20 +41,92 @@ class Linear(Module):
...
-        return f"input_dims={self.weight.shape[1]}, output_dims={self.weight.shape[0]}, bias={'bias' in self}"
+        return (
+            f"input1_dims={self.input1_dims}, input2_dims={self.input2_dims}, output_dims={self.output_dims}, "
+            f"bias={'bias' in self}"
+        )
 
-    def __call__(self, x):
-        x = x @ self.weight.T
+    def __call__(self, input1, input2):
+        output = (input1 @ self.weight * input2.reshape(1, *input2.shape)).sum(-1).T
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:129]: This should be done with two matrix multiplications not a multiplication + sum.

File: python/mlx/nn/layers/linear.py [hunk 5]
```diff
@@ -26,20 +41,92 @@ class Linear(Module):
...
             high=scale,
-            shape=(output_dims, input_dims),
+            shape=(self.output_dims, self.input_dims),
         )
+        if "bias" in self:
+            self.bias = mx.random.uniform(
+                low=-scale,
+                high=scale,
+                shape=(self.output_dims,),
+            )
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:64]: We should consider a more general solution to reseting parameters rather than just adding this to linear. For simplicity, I would remove it from this PR and if you are up for exploring a more general solution, let's put it into a follow on PR.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (awni) [inline python/mlx/nn/layers/linear.py:83]: Also you need some kind of indexing on the y and W for this to make sense. Like:

$y_i = x_1^\top W_i x_2 + b_i$

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for adding this! 🚀 

I left a few comments / questions. Please address, then we can merge!

File: python/mlx/nn/layers/linear.py (hunk 1 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py:9]: Here are some use cases for the identity layer:
1. Removing parts of existing neural networks:
- _Example_: Directly obtain features instead of class logits from pretrained network.
```python
model = models.resnet18()
model.fc = nn.Identity() # Replace final linear layer with Identity
features = model(input_tensor)
```
2. Conditional Computation:
- Enables different computations based on conditions while keeping model structure.
- _Example_: batchnorm as condition
```python
batch_norm = nn.BatchNorm2d
if dont_use_batch_norm:
    batch_norm = Identity

nn.Sequential(
    ...
    batch_norm(N, momentum=0.05),
    ...
)
```
References:
- https://github.com/pytorch/pytorch/issues/9160#issuecomment-456766137
- https://github.com/pytorch/pytorch/pull/19249
- https://discuss.pytorch.org/t/what-is-nn-identity-used-for/125457/14

File: python/mlx/nn/layers/linear.py (hunk 2 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py:45]: I thought keeping this variables would make debugging and reading the code easier. I'll be changing it to use `self.weight.shape` in the next commit.

Comment (awni) [reply python/mlx/nn/layers/linear.py:45]: Actually I think it's ok to keep it if it's needed for `reset_parameters()`, but let's include it in a follow on PR when we add that.

File: python/mlx/nn/layers/linear.py (hunk 1 above)
Comment (awni) [reply python/mlx/nn/layers/linear.py:9]: Makes sense, thanks for the input. I like the first use case, that's clever.

File: python/mlx/nn/layers/linear.py (hunk 2 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py:45]: What do you think should be included in the initialization? I think using 'reset_parameters' is important for any solution. It makes it easy to set up weights, especially for those who prefer a simple approach without specific requirements

Comment (awni) [reply python/mlx/nn/layers/linear.py:45]: I'm not sure I understand your question. 

I agree we should have a `reset_parameters` (or some equivalent functionality). My only point is that it should be module general rather than specific to the linear layer. That way you could do something like `module.reset_parameters()`

Comment (awni) [reply python/mlx/nn/layers/linear.py:45]: For that to work, we probably need an initialize function or something like that to be part of the base Module class? Idk I haven't thought about it much. Might be worth looking at how other frameworks do it

Comment (nuniz) [reply python/mlx/nn/layers/linear.py:45]: I agree with you. I was addressing the case where you want to initialize specific layers while leaving others. We can leave it for future work.

File: python/mlx/nn/layers/linear.py (hunk 4 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py:129]: Hi, I think the math is ok [(reference)](https://www.quora.com/What-is-a-bilinear-tensor-layer-in-contrast-to-a-standard-linear-neural-network-layer-or-how-can-I-imagine-it) and the implementation aligns with earlier versions of torch (https://github.com/pytorch/pytorch/pull/1146, https://github.com/pytorch/pytorch/pull/1104).

I've implemented several versions and compared them to `torch.nn.functional.bilinear`. To avoid using a for loop, I created a custom implementation that could be replaced with `einsum` in the future.

Let me know your thoughts and if I missed something.

```python
import torch
import torch.nn.functional as F

# Set deterministic algorithms for reproducibility
torch.use_deterministic_algorithms(True)
torch.manual_seed(0)
torch.backends.cudnn.deterministic = True

# Generate random tensors
x1 = torch.randn((10, 20))
x2 = torch.randn((10, 40))
w = torch.randn((60, 20, 40))

# Original implementation using F.bilinear
res_original = F.bilinear(x1, x2, w)

# Naive implementation
res_naive= torch.cat([(x1.mm(w[i]).mul(x2)).sum(1, keepdims=True) for i in range(w.shape[0])], 1)

# Einsum implementation
res_einsum = torch.einsum('bi,kij,bj->bk', x1, w, x2)

# Custom implementation
res_custom= (((torch.matmul(x1, w) * x2.unsqueeze(0))).sum(-1)).T

# Print mean absolute differences between implementations
print('Mean Absolute Difference - Naive:', abs(res_original - res_naive).mean())
print('Mean Absolute Difference - Einsum:', abs(res_original - res_einsum).mean())
print('Mean Absolute Difference - Custom:', abs(res_original - res_custom).mean())
```

Output:
```python
Mean Absolute Difference - Naive: tensor(2.7407e-06)
Mean Absolute Difference - Einsum: tensor(6.3483e-06)
Mean Absolute Difference - Custom: tensor(2.7407e-06)
```

**Comment by awni** [comment]
@nuniz let me know when you'd like me to review this again! I'm looking forward to merging it!

File: python/mlx/nn/layers/linear.py (hunk 2 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: Got it. I've reverted the changes for now, and we'll work on that in a future PR.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: I've changed the math to x_1 and x_2 https://github.com/ml-explore/mlx/pull/315/commits/a4b43db9af5a00285213d7c9ae710166663aca58 (Line 80)

File: python/mlx/nn/layers/linear.py (hunk 5 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: Got it. I've reverted the changes for now, and we'll work on that in a future PR. https://github.com/ml-explore/mlx/pull/315/commits/a4b43db9af5a00285213d7c9ae710166663aca58.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: I'm not sure if we really need indexing because the Linear layer doesn't use indexing either. Let me know what you think

**Comment by nuniz** [comment]
Hi @awni I've made the necessary changes. You can review them again and please relate to my comments about the Bilinear layer.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (awni) [reply python/mlx/nn/layers/linear.py]: Ok fair enough!

File: python/mlx/nn/layers/linear.py (hunk 4 above)
Comment (awni) [reply python/mlx/nn/layers/linear.py]: Right I did not mean to imply it was in pro. Just that it will be a lot faster to use two matrix multiplications instead of a multiply and a reduce. Can you switch it?

Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: If I understand correctly, you're asking for a bilinear transformation using two matrix multiplications. However, this leads to extra calculations and a longer runtime due to the bilinear operation being only the diagonal of `x1 @ w @ x2.T`. I've removed unnecessary reshaping in the latest commit. Until we have the einsum operation, I believe the method I've used is the most efficient. Let me know if I missed anything.

Comment (awni) [reply python/mlx/nn/layers/linear.py]: Assuming the inputs have the dimensions expanded appropriately:

- `x1` has shape `[B, 1, 1, D1]`
- `W` has shape `[1, D, D1, D2]`
- `x2` has shape `[B, 1, D2]`

You can do something like:

```
y = mx.squeeze(x1 @ w, -2) # shape B x D x D2
z = x2 @ y.swapaxes(-1, -2) # shape B x 1 x D
return mx.squeeze(z, -2)
```

You might play around with it and find a simpler implementation.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (awni) [reply python/mlx/nn/layers/linear.py]: On second thought, I think this computation is really different than the linear and subscripts make sense here. It makes it more clear that it's not a single bilinear matrix but $d$ bilinear matrices. For example, in the equation $y = x W$ the output $y$ has dimension $d$. But here the output $y = x_1^\top W x_2$ has dimension $1$ which is not right.

File: python/mlx/nn/layers/linear.py (hunk 4 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: Thanks for writing the implementation; I haven't thought of it. I made a runtime comparison using `timeit` with the conditions `B=100, D=60, D1=20, D2=40`, and my implementation was >20% faster. Let me know if you still want me to change this.

Comment (awni) [reply python/mlx/nn/layers/linear.py]: > and my implementation was >20% faster. Let me know if you still want me to change this.

Can you post the benchmark? That doesn't make sense that the broadcast + reduce is faster. 

It's possible for really small `D*` it is comparable but for large `D*` using a matmul should be significantly faster and also use way less memory.

File: python/mlx/nn/layers/linear.py (hunk 3 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: Yeah, each multiplication results in a scalar, but I think its similar to `Linear` because it corresponding to a `d` linear "vectors". In any case, I've changed the math formula to include 'i' in the latest commit  https://github.com/ml-explore/mlx/pull/315/commits/a9512835611dbaee74f567d9c6d2add34c0eb306.

File: python/mlx/nn/layers/linear.py (hunk 4 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: Hi, Thanks for catching that! I took another test, and you're right. I've made the changes you asked for in [this commit](https://github.com/ml-explore/mlx/pull/315/commits/8e22ab1e5a3ebb8e79ed4273dac9273b7e223bb6). Let me know if everything looks good now.

File: python/mlx/nn/layers/linear.py [hunk 6]
```diff
@@ -15,31 +31,98 @@ class Linear(Module):
...
+
+    def _extra_repr(self) -> str:
+        return (
+            f"input1_dims={self.weight.shape[2]}, input2_dims={self.weight.shape[3]}, "
+            f"output_dims={self.weight.shape[1]}, bias={'bias' in self}"
+        )
+
+    def __call__(self, x1: mx.array, x2: mx.array) -> mx.array:
+        x1 = x1.reshape(x1.shape[0], 1, 1, x1.shape[1])
+        x2 = x2.reshape(x2.shape[0], 1, x2.shape[1])
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:123]: Could you use `mx.expand_dims` here and negative indexing so that we don't constrain the inputs to be 2 dimensional.

File: python/mlx/nn/layers/linear.py [hunk 7]
```diff
@@ -15,31 +31,98 @@ class Linear(Module):
...
+
+        y_i = x_1^\top W_i x_2 + b_i
+
+    where
+    :math:`W` has shape ``[output_dims, input1_dims, input2_dims]``.
+    :math:`b` has shape ``[output_dims, ]``.
+    :math:`i` is the index for output dimensions.
+
+    The values are initialized from :math:`\mathcal{U}(-{k}, {k})`, where
+    :math:`k = \frac{1}{\sqrt{input1\_dims}}`.
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:88]:
```suggestion
    where  :math:`W` has shape ``[output_dims, input1_dims, input2_dims]``, 
    :math:`b` has shape ``[output_dims ]``, and :math:`i` indexes the
    output dimension.

    The values are initialized from the uniform distribution :math:`\mathcal{U}(-{k}, {k})`,
    where :math:`k = \frac{1}{\sqrt{D_1}}` and :math:`D_1` is ``input1_dims``.
```

File: python/mlx/nn/layers/linear.py [hunk 8]
```diff
@@ -15,31 +31,98 @@ class Linear(Module):
 
         y = W^\top x + b
 
-    where :math:`W` has shape ``[output_dims, input_dims]``.
+    where:
+    :math:`W` has shape ``[output_dims, input_dims]``.
+    :math:`b` has shape ``[output_dims, ]``.
+
+    The values are initialized from :math:`\mathcal{U}(-{k}, {k})`, where
+    :math:`k = \frac{1}{\sqrt{input\_dims}}`.
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:39]:
```suggestion
    where :math:`W` has shape ``[output_dims, input_dims]`` and
    :math:`b` has shape ``[output_dims]``.

    The values are initialized from the uniform distribution :math:`\mathcal{U}(-{k}, {k})`,
    where :math:`k = \frac{1}{\sqrt{D_i}}` and :math:`D_i` is equal to ``input_dims``.
```

File: python/mlx/nn/layers/linear.py [hunk 9]
```diff
@@ -15,31 +31,98 @@ class Linear(Module):
 
         y = W^\top x + b
 
-    where :math:`W` has shape ``[output_dims, input_dims]``.
+    where:
```
Comment (awni) [inline python/mlx/nn/layers/linear.py:34]: Something doesn't add up between the equation and the shape of `W`. I think the equation should read $x W^\top + b$ instead.

**Review by awni — CHANGES_REQUESTED** [review]
Looks great! A few more nits, then we can merge it in 🚀

File: python/mlx/nn/layers/linear.py (hunk 7 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: I've made the change

File: python/mlx/nn/layers/linear.py (hunk 8 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: I've made the change

File: python/mlx/nn/layers/linear.py (hunk 9 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py:34]: I've changed it to `y = x W^\top + b`

File: python/mlx/nn/layers/linear.py (hunk 6 above)
Comment (nuniz) [reply python/mlx/nn/layers/linear.py]: That's smart, I've changed it to:
`x1 = mx.expand_dims(x1, axis=(-2, -3))`
`x2 = mx.expand_dims(x2, axis=(-2))`

**Review by awni — APPROVED** [review]

**Comment by awni** [comment]
Thanks for adding this!!
